salt/modules/nxos_upgrade.py

```diff
@@ -0,0 +1,405 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018 Cisco and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+'''
+Execution module to upgrade Cisco NX-OS Switches.
+
+.. versionadded:: xxxx.xx.x
+
+This module supports execution using a Proxy Minion or Native Minion:
+1) Proxy Minion: Connect over SSH or NX-API HTTP(S).
+   See :mod:`salt.proxy.nxos <salt.proxy.nxos>` for proxy minion setup details.
+2) Native Minion: Connect over NX-API Unix Domain Socket (UDS).
+   Install the minion inside the GuestShell running on the NX-OS device.
+
+:maturity:   new
+:platform:   nxos
+:codeauthor: Michael G Wiebe
+
+.. note::
+
+    To use this module over remote NX-API the feature must be enabled on the
+    NX-OS device by executing ``feature nxapi`` in configuration mode.
+
+    This is not required for NX-API over UDS.
+
+    Configuration example:
+
+    .. code-block:: bash
+
+        switch# conf t
+        switch(config)# feature nxapi
+
+    To check that NX-API is properly enabled, execute ``show nxapi``.
+
+    Output example:
+
+    .. code-block:: bash
+
+        switch# show nxapi
+        nxapi enabled
+        HTTPS Listen on port 443
+'''
+from __future__ import absolute_import, print_function, unicode_literals
+
+# Import python stdlib
+import logging
+import re
+import ast
+import time
+from salt.ext.six import string_types
+
+# Import Salt libs
+from salt.exceptions import (NxosError, CommandExecutionError)
+from salt.ext.six.moves import range
+
+__virtualname__ = 'nxos'
```

Comment:
This is going to conflict with the already existing module: https://docs.saltstack.com/en/develop/ref/modules/all/salt.modules.nxos.html

I'd recommend to rather enhance one of the existing `nxos` or [`nxos_api`](https://docs.saltstack.com/en/develop/ref/proxy/all/salt.proxy.nxos_api.html) modules instead, by adding these functions.